### PR TITLE
Use new API of android gradle plugin 3.3.0 for Task Configuration Avo…

### PR DESCRIPTION
**Pull request by bastienpaulfr**
…idance

Dear @vanniktech 

Could you please review these changes that are addressing #65 

Many thanks

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -20,6 +20,7 @@ apply plugin: 'com.github.ben-manes.versions'
 apply plugin: 'com.gradle.plugin-publish'
 apply plugin: 'com.vanniktech.android.junit.jacoco'
 apply plugin: "com.vanniktech.maven.publish"
+apply plugin: 'maven-publish'
```
Comment (vanniktech) [inline build.gradle]: Remove this please. You can install the plugin locally via `installArchives` task

File: build.gradle [hunk 2]
```diff
@@ -62,6 +63,24 @@ pluginBundle {
 }
 
 task wrapper(type: Wrapper) {
-  gradleVersion = '4.8'
+  gradleVersion = '4.10.1'
+  //noinspection UnnecessaryQualifiedReference
```
Comment (vanniktech) [inline build.gradle]: Remove this please

File: build.gradle [hunk 3]
```diff
@@ -62,6 +63,24 @@ pluginBundle {
...
   distributionType = Wrapper.DistributionType.ALL
 }
+
+// Using :publishToMavenLocal will put the jar in $home/.m2 repo
```
Comment (vanniktech) [inline build.gradle]: Remove this please (see comment from above)

File: build.gradle [hunk 4]
```diff
@@ -62,6 +63,24 @@ pluginBundle {
...
+      groupId = GROUP
+      artifactId = POM_ARTIFACT_ID
+
+      //artifact jar
+      from components.java
+    }
+  }
+}
+
+tasks.named("publishToMavenLocal") .configure {
```
Comment (vanniktech) [inline build.gradle]: Remove this too

File: src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy [hunk 5]
```diff
@@ -98,37 +109,43 @@ class Generation implements Plugin<Project> {
...
         String taskName = genJavadocTaskName(project, variant)
 
         logger.debug "Create task ${taskName} for project ${project.name}, variant ${variant.name}"
         // We are creating the task with the same name only once. This is a protection if task already exists
         // because the name of the task is configurable and can be the same for several variants.
-        if (project.tasks.findByPath(taskName)) {
+        if (findTask(project, taskName)) {
             logger.debug "task $taskName already exists"
-            return project.tasks.findByPath(taskName)
+            //return findTask(project, taskName)
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy]: What's up with that? Can this be removed or why is it still here?

File: src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy [hunk 6]
```diff
@@ -98,37 +109,43 @@ class Generation implements Plugin<Project> {
...
-        project.task(taskName, type: Javadoc) {
+    private TaskProvider createJavadocTask(final Project project, variant, String taskName) {
+        project.tasks.register(taskName, Javadoc) {
             title = "Documentation for ${project.name} at version ${project.android.defaultConfig.versionName}"
             description = "Generates javadoc for $variant.name variant."
             group = 'Documentation'
 
             destinationDir = getJavadocFolder(project, variant)
-            source = variant.javaCompiler.source
+            source = variant.sourceSets.collect { it.java.sourceFiles }.inject { m, i -> m + i }
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy]: Why was this changed?

File: src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy [hunk 7]
```diff
@@ -158,12 +175,15 @@ class Generation implements Plugin<Project> {
 
     private void addDeleteJavadocTaskToCleanTaskIn(final Project project, variant) {
         String taskName = genDeleteTaskName(project, variant)
-        if (project.tasks.findByPath(taskName)) {
+        if (findTask(project, taskName)) {
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy]: This is inconsistent with `createRootTask`. How come? I'd expect to not use `findTask` here or also use `findTask` in `createRootTask`.

File: src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy [hunk 8]
```diff
@@ -186,4 +206,15 @@ class Generation implements Plugin<Project> {
     private static File getJavadocFolder(final Project project, variant) {
         return new File("${project.androidJavadoc.outputDir(project)}", "${project.androidJavadoc.taskNameTransformer(variant)}")
     }
+
+    @Nullable
+    private TaskProvider findTask(final Project project, String taskName) {
+        def ret = null
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy]: `taskProvider`

**Review by vanniktech — COMMENTED** [review]
Some questions but overall looks good.

File: src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy (hunk 7 above)
Comment (bastienpaulfr) [reply src/main/groovy/com/vanniktech/android/javadoc/Generation.groovy]: findTask is just a helper that uses new TaskAvoidance API (`project.tasks.named()`)

I am using this to avoid creating a task that already exists for some reason i don't know

**Review by vanniktech — APPROVED** [review]
Thanks for making the changes!
